src/tests/netsim_tests.rs

```diff
@@ -0,0 +1,389 @@
+            SubnetV4::global(),
+            netsim::node::router_v4((server_node, client_node))
+        );
+
+        spawn_complete
+        .resume_unwind()
+        .map(|((), ())| ())
+    }));
+    res.void_unwrap()
+}
+
+#[test]
+fn rendezvous_connect_over_poor_connection() {
+    let _ = env_logger::init();
+
+    let mut core = unwrap!(Core::new());
+    let handle = core.handle();
+
+    let send_data_a = util::random_vec(1024);
+    let send_data_a_clone = send_data_a.clone();
+    let send_data_b = util::random_vec(1024);
+    let send_data_b_clone = send_data_b.clone();
+
+    let (drop_tx_a0, drop_rx_a0) = future_utils::drop_notify();
+    let (drop_tx_b0, drop_rx_b0) = future_utils::drop_notify();
+    let (addr_tx_a0, addr_rx_a0) = oneshot::channel();
+    let (addr_tx_b0, addr_rx_b0) = oneshot::channel();
+    let (drop_tx_a1, drop_rx_a1) = future_utils::drop_notify();
+    let (drop_tx_b1, drop_rx_b1) = future_utils::drop_notify();
+    let (addr_tx_a1, addr_rx_a1) = oneshot::channel();
+    let (addr_tx_b1, addr_rx_b1) = oneshot::channel();
+    let (ci_tx_a, ci_rx_a) = oneshot::channel();
+    let (ci_tx_b, ci_rx_b) = oneshot::channel();
+    let res = core.run(future::lazy(|| {
+        let rendezvous_server_node_0 = netsim::node::endpoint_v4(|ip| {
+            let mut core = unwrap!(Core::new());
+            let handle = core.handle();
+
+            let res = core.run(future::lazy(|| {
+                let config = unwrap!(ConfigFile::new_temporary());
+                let addr = SocketAddr::V4(SocketAddrV4::new(ip, 1234));
+                unwrap!(config.write()).listen_addresses = vec![PaAddr::Utp(addr)];
+                Service::with_config(&handle, config, util::random_id())
+                .map_err(|e| panic!("error creating service: {}", e))
+                .and_then(move |service| {
+                    let server_info = PeerInfo {
+                        addr: PaAddr::Utp(SocketAddr::V4(SocketAddrV4::new(ip, 1234))),
+                        pub_key: service.public_key(),
+                    };
+
+                    unwrap!(addr_tx_a0.send(server_info.clone()));
+                    unwrap!(addr_tx_b0.send(server_info));
+
+                    let listening = {
+                        service
+                        .start_listening()
+                        .map_err(|e| panic!("error starting listeners: {}", e))
+                        .collect()
+                    };
+
+                    drop_rx_a0
+                    .and_then(|()| {
+                        drop_rx_b0
+                        .map(|()| {
+                            drop(service);
+                        })
+                    })
+                    .while_driving(listening)
+                    .map_err(|(e, _)| e)
+                    .map(|((), _listening)| ())
+                })
+            }));
+            res.void_unwrap()
+        });
+
+        let rendezvous_server_node_1 = netsim::node::endpoint_v4(|ip| {
```

Comment:
Why do we need 2 rendezvous nodes?

Reply:
Why not? I wouldn't have found that bug yesterday without two rendezvous nodes. It makes the test a bit more realistic.

Reply:
Hm, maybe, when I do some manual testing, I only have one rendezvous peer running, which seems to have no difference since all it does is return peer's address.
My motivation for having only 1 rendezvous peer is simplicity - less code :)

Reply:
Yeah okay that's a good motivation :)

The purpose of these netsim tests though is to be as realistic as possible so that we'll know whether the code works in real life.